docs/spec.md

```diff
@@ -44,6 +44,56 @@ If a method called with incorrect types the behaviour is unspecified.
 Also if a method accepts a function it must only apply the function in accordance with
 the type signature (i.e. provide the correct number of arguments of the correct types).
 
+
+## Parametricity
+
+All methods' implementations should only use type information about arguments that is known from the
+method's type signature. It's not allowed to inspect arguments or values that they produce to get
+more information about their types. In other words they should be [parametrically polymorphic](https://en.wikipedia.org/wiki/Parametric_polymorphism).
+
+For example let's take a look at Functor's `map` signature:
+
+```
+map :: Functor f => Type f ~> (a → b, f a) → f b
+```
+
+There are three type variables in it: `f`, `a`, and `b`. Also we have some restrictions:
+
+  1. `Functor f` says that `f a` is a value of a Functor.
+  2. `Type f ~>` means that we're writing implementation of `map`
+     for a [type dictionary](#type) `Type f`. In this case we know what specific
+     types `Type f` works with, so we know everything about `f`.
+
+We don't have any restrictions for types `a` and `b`, so we don't know anything about them.
+And we're not allowed to inspect them.
+
+Here's an example of a correct `map` implementation that works with a simple wrapper `{x: ...}`:
+
+```js
+X.map = (f, v) => {
+  const a = v.x
+  const b = f(a)
+  return {x: b}
+}
+```
+
+And here's one that violates parametricity requirement although fits into type signature otherwise:
```

Comment:
Also the less contrived I'm trying to make it the bigger it grows :)

I think the purpose of example here should be not to explain **why** we require parametricity, but rather to explain **what** exactly we require. Answers to **why** questions are very important but should not live in a specification, imo.

So maybe current example is good enough, or we could switch to a Maybe example but leave only first code snippet, with an additional code comment:

```js
  if (b === undefined) { // this is not allowed
    return Maybe.Nothing
  }
```